

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>gnomad.utils.liftover &mdash; gnomad master documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> gnomad
          

          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../examples/index.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api_reference/index.html">API Reference</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">gnomad</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>gnomad.utils.liftover</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for gnomad.utils.liftover</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">os.path</span> <span class="kn">import</span> <span class="n">basename</span><span class="p">,</span> <span class="n">dirname</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Union</span>

<span class="kn">import</span> <span class="nn">hail</span> <span class="k">as</span> <span class="nn">hl</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.reference_genome</span> <span class="kn">import</span> <span class="n">get_reference_genome</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span>
    <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(asctime)s</span><span class="s2"> (</span><span class="si">%(name)s</span><span class="s2"> </span><span class="si">%(lineno)s</span><span class="s2">): </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="n">datefmt</span><span class="o">=</span><span class="s2">&quot;%m/</span><span class="si">%d</span><span class="s2">/%Y %I:%M:%S %p&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>


<div class="viewcode-block" id="get_checkpoint_path"><a class="viewcode-back" href="../../../api_reference/utils/liftover.html#gnomad.utils.liftover.get_checkpoint_path">[docs]</a><span class="k">def</span> <span class="nf">get_checkpoint_path</span><span class="p">(</span><span class="n">gnomad</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">is_table</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a checkpoint path for Table</span>

<span class="sd">    :param gnomad: Whether data is gnomAD data</span>
<span class="sd">    :param data_type: Data type (exomes or genomes for gnomAD; not used otherwise)</span>
<span class="sd">    :param path: Path to input Table/MatrixTable (if data is not gnomAD data)</span>
<span class="sd">    :param is_table: Whether data is a Table</span>
<span class="sd">    :return: Output checkpoint path</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">gnomad</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;gs://gnomad/liftover/release/2.1.1/ht/</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">/gnomad.</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">.r2.1.1.liftover.ht&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">out_name</span> <span class="o">=</span> <span class="n">basename</span><span class="p">(</span><span class="n">path</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">out</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">dirname</span><span class="p">(</span><span class="n">path</span><span class="p">)</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">out_name</span><span class="si">}</span><span class="s2">_lift&quot;</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out</span><span class="si">}</span><span class="s2">.ht&quot;</span> <span class="k">if</span> <span class="n">is_table</span> <span class="k">else</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out</span><span class="si">}</span><span class="s2">.mt&quot;</span></div>


<div class="viewcode-block" id="get_liftover_genome"><a class="viewcode-back" href="../../../api_reference/utils/liftover.html#gnomad.utils.liftover.get_liftover_genome">[docs]</a><span class="k">def</span> <span class="nf">get_liftover_genome</span><span class="p">(</span><span class="n">t</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Infers genome build of input data and assumes destination build. Prepares to liftover to destination genome build</span>

<span class="sd">    :param t: Input Table or MatrixTable</span>
<span class="sd">    :return: List of source build (with liftover chain added) and destination build (with sequence loaded)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Inferring build of input&quot;</span><span class="p">)</span>
    <span class="n">build</span> <span class="o">=</span> <span class="n">get_reference_genome</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">locus</span><span class="p">)</span><span class="o">.</span><span class="n">name</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Loading reference genomes, adding chain file, and loading fasta sequence for destination build&quot;</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="n">build</span> <span class="o">==</span> <span class="s2">&quot;GRCh38&quot;</span><span class="p">:</span>
        <span class="n">source</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">get_reference</span><span class="p">(</span><span class="s2">&quot;GRCh38&quot;</span><span class="p">)</span>
        <span class="n">target</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">get_reference</span><span class="p">(</span><span class="s2">&quot;GRCh37&quot;</span><span class="p">)</span>
        <span class="n">chain</span> <span class="o">=</span> <span class="s2">&quot;gs://hail-common/references/grch38_to_grch37.over.chain.gz&quot;</span>
        <span class="n">target</span><span class="o">.</span><span class="n">add_sequence</span><span class="p">(</span>
            <span class="s2">&quot;gs://hail-common/references/human_g1k_v37.fasta.gz&quot;</span><span class="p">,</span>
            <span class="s2">&quot;gs://hail-common/references/human_g1k_v37.fasta.fai&quot;</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">source</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">get_reference</span><span class="p">(</span><span class="s2">&quot;GRCh37&quot;</span><span class="p">)</span>
        <span class="n">target</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">get_reference</span><span class="p">(</span><span class="s2">&quot;GRCh38&quot;</span><span class="p">)</span>
        <span class="n">chain</span> <span class="o">=</span> <span class="s2">&quot;gs://hail-common/references/grch37_to_grch38.over.chain.gz&quot;</span>
        <span class="n">target</span><span class="o">.</span><span class="n">add_sequence</span><span class="p">(</span>
            <span class="s2">&quot;gs://hail-common/references/Homo_sapiens_assembly38.fasta.gz&quot;</span><span class="p">,</span>
            <span class="s2">&quot;gs://hail-common/references/Homo_sapiens_assembly38.fasta.fai&quot;</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="n">source</span><span class="o">.</span><span class="n">add_liftover</span><span class="p">(</span><span class="n">chain</span><span class="p">,</span> <span class="n">target</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">source</span><span class="p">,</span> <span class="n">target</span><span class="p">]</span></div>


<div class="viewcode-block" id="lift_data"><a class="viewcode-back" href="../../../api_reference/utils/liftover.html#gnomad.utils.liftover.lift_data">[docs]</a><span class="k">def</span> <span class="nf">lift_data</span><span class="p">(</span>
    <span class="n">t</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">],</span>
    <span class="n">gnomad</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">rg</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">genetics</span><span class="o">.</span><span class="n">ReferenceGenome</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Lifts input Table or MatrixTable from one reference build to another</span>

<span class="sd">    :param t: Table or MatrixTable</span>
<span class="sd">    :param gnomad: Whether data is gnomAD data</span>
<span class="sd">    :param data_type: Data type (exomes or genomes for gnomAD; not used otherwise)</span>
<span class="sd">    :param path: Path to input Table/MatrixTable (if data is not gnomAD data)</span>
<span class="sd">    :param rg: Reference genome</span>
<span class="sd">    :param overwrite: Whether to overwrite data</span>
<span class="sd">    :return: Table or MatrixTablewith liftover annotations</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Annotating input with liftover coordinates&quot;</span><span class="p">)</span>
    <span class="n">liftover_expr</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;new_locus&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">liftover</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span> <span class="n">rg</span><span class="p">,</span> <span class="n">include_strand</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
        <span class="s2">&quot;old_locus&quot;</span><span class="p">:</span> <span class="n">t</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">t</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">t</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="o">**</span><span class="n">liftover_expr</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span>
        <span class="k">else</span> <span class="n">t</span><span class="o">.</span><span class="n">annotate_rows</span><span class="p">(</span><span class="o">**</span><span class="n">liftover_expr</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="n">no_target_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_missing</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">new_locus</span><span class="p">))</span>
    <span class="n">num_no_target</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">t</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="n">no_target_expr</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span>
        <span class="k">else</span> <span class="n">t</span><span class="o">.</span><span class="n">aggregate_rows</span><span class="p">(</span><span class="n">no_target_expr</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Filtering out </span><span class="si">{</span><span class="n">num_no_target</span><span class="si">}</span><span class="s2"> sites that failed to liftover&quot;</span><span class="p">)</span>
    <span class="n">keep_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">new_locus</span><span class="p">)</span>
    <span class="n">t</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">keep_expr</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="k">else</span> <span class="n">t</span><span class="o">.</span><span class="n">filter_rows</span><span class="p">(</span><span class="n">keep_expr</span><span class="p">)</span>

    <span class="n">row_key_expr</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">t</span><span class="o">.</span><span class="n">new_locus</span><span class="o">.</span><span class="n">result</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="n">t</span><span class="o">.</span><span class="n">alleles</span><span class="p">}</span>
    <span class="n">t</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">t</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="o">**</span><span class="n">row_key_expr</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span>
        <span class="k">else</span> <span class="n">t</span><span class="o">.</span><span class="n">key_rows_by</span><span class="p">(</span><span class="o">**</span><span class="n">row_key_expr</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Writing out lifted over data&quot;</span><span class="p">)</span>
    <span class="n">t</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span>
        <span class="n">get_checkpoint_path</span><span class="p">(</span><span class="n">gnomad</span><span class="p">,</span> <span class="n">data_type</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)),</span>
        <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">t</span></div>


<div class="viewcode-block" id="annotate_snp_mismatch"><a class="viewcode-back" href="../../../api_reference/utils/liftover.html#gnomad.utils.liftover.annotate_snp_mismatch">[docs]</a><span class="k">def</span> <span class="nf">annotate_snp_mismatch</span><span class="p">(</span>
    <span class="n">t</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">],</span> <span class="n">rg</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">genetics</span><span class="o">.</span><span class="n">ReferenceGenome</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Annotates mismatches between reference allele and allele in reference fasta</span>

<span class="sd">    Assumes input Table/MatrixTable has t.new_locus annotation</span>

<span class="sd">    :param t: Table/MatrixTable of SNPs to be annotated</span>
<span class="sd">    :param rg: Reference genome with fasta sequence loaded</span>
<span class="sd">    :return: Table annotated with mismatches between reference allele and allele in fasta</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Filtering to SNPs&quot;</span><span class="p">)</span>
    <span class="n">snp_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_snp</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">t</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">t</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">snp_expr</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="k">else</span> <span class="n">t</span><span class="o">.</span><span class="n">filter_rows</span><span class="p">(</span><span class="n">snp_expr</span><span class="p">)</span>

    <span class="n">mismatch_expr</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;reference_mismatch&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">cond</span><span class="p">(</span>
            <span class="n">t</span><span class="o">.</span><span class="n">new_locus</span><span class="o">.</span><span class="n">is_negative_strand</span><span class="p">,</span>
            <span class="p">(</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                <span class="o">!=</span> <span class="n">hl</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span>
                    <span class="n">t</span><span class="o">.</span><span class="n">locus</span><span class="o">.</span><span class="n">contig</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">locus</span><span class="o">.</span><span class="n">position</span><span class="p">,</span> <span class="n">reference_genome</span><span class="o">=</span><span class="n">rg</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="p">(</span>
                <span class="n">t</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="o">!=</span> <span class="n">hl</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span>
                    <span class="n">t</span><span class="o">.</span><span class="n">locus</span><span class="o">.</span><span class="n">contig</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">locus</span><span class="o">.</span><span class="n">position</span><span class="p">,</span> <span class="n">reference_genome</span><span class="o">=</span><span class="n">rg</span>
                <span class="p">)</span>
            <span class="p">),</span>
        <span class="p">)</span>
    <span class="p">}</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Checking if reference allele matches what is in reference fasta&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;For SNPs on the negative strand, make sure the reverse complement of the ref alleles matches what is in the ref fasta&quot;</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="p">(</span>
        <span class="n">t</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="o">**</span><span class="n">mismatch_expr</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span>
        <span class="k">else</span> <span class="n">t</span><span class="o">.</span><span class="n">annotate_rows</span><span class="p">(</span><span class="o">**</span><span class="n">mismatch_expr</span><span class="p">)</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="check_mismatch"><a class="viewcode-back" href="../../../api_reference/utils/liftover.html#gnomad.utils.liftover.check_mismatch">[docs]</a><span class="k">def</span> <span class="nf">check_mismatch</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">expressions</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Checks for mismatches between reference allele and allele in reference fasta</span>

<span class="sd">    :param ht: Table to be checked</span>
<span class="sd">    :return: StructExpression containing counts for mismatches and count for all variants on negative strand</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">mismatch</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">total_variants</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count</span><span class="p">(),</span>
            <span class="n">total_mismatch</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">reference_mismatch</span><span class="p">),</span>
            <span class="n">negative_strand</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">new_locus</span><span class="o">.</span><span class="n">is_negative_strand</span><span class="p">),</span>
            <span class="n">negative_strand_mismatch</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span>
                <span class="n">ht</span><span class="o">.</span><span class="n">new_locus</span><span class="o">.</span><span class="n">is_negative_strand</span> <span class="o">&amp;</span> <span class="n">ht</span><span class="o">.</span><span class="n">reference_mismatch</span>
            <span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">mismatch</span></div>


<div class="viewcode-block" id="liftover_using_gnomad_map"><a class="viewcode-back" href="../../../api_reference/utils/liftover.html#gnomad.utils.liftover.liftover_using_gnomad_map">[docs]</a><span class="k">def</span> <span class="nf">liftover_using_gnomad_map</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="n">data_type</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Liftover a gnomAD table using already-established liftover file. Warning: shuffles!</span>

<span class="sd">    :param ht: Input Hail table</span>
<span class="sd">    :param data_type: one of &quot;exomes&quot; or &quot;genomes&quot; which to map across</span>
<span class="sd">    :return: Lifted over table</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span> <span class="nn">gnomad.resources.grch37.gnomad</span> <span class="kn">import</span> <span class="n">liftover</span>

    <span class="n">lift_ht</span> <span class="o">=</span> <span class="n">liftover</span><span class="p">(</span><span class="n">data_type</span><span class="p">)</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="n">original_locus</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span> <span class="n">original_alleles</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span>
        <span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">lift_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="o">**</span><span class="n">ht</span><span class="p">[(</span><span class="n">lift_ht</span><span class="o">.</span><span class="n">original_locus</span><span class="p">,</span> <span class="n">lift_ht</span><span class="o">.</span><span class="n">original_alleles</span><span class="p">)]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">)</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>